larpandoracontent/LArHelpers/LArDiscreteProbabilityHelper.cc

```diff
@@ -0,0 +1,130 @@
+            LArDiscreteProbabilityHelper::MakeRandomisedSample(t1,randomNumberGenerator),
+            LArDiscreteProbabilityHelper::MakeRandomisedSample(t2,randomNumberGenerator)));
+
+        if ((rRandomised-rNominal) > std::numeric_limits<float>::epsilon())
+            nExtreme++;
+    }
+
+    return static_cast<float>(nExtreme)/static_cast<float>(nPermutations);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+template <typename T>
+float LArDiscreteProbabilityHelper::CalculateCorrelationCoefficientPValueFromStudentTDistribution(const T &t1, 
+    const T &t2, const size_t nIntegrationSteps)
+{
+    const float correlation(LArDiscreteProbabilityHelper::CalculateCorrelationCoefficient(t1,t2));
+    const float dof(static_cast<float>(LArDiscreteProbabilityHelper::GetSize(t1)) - 2.f);
+    const float tTestStatistic(correlation*sqrt(dof)/(sqrt(1.f - correlation*correlation)));
+    const float tDistCoeff(std::tgamma(0.5f*(dof + 1.f)) / std::tgamma(0.5f*dof)/(std::sqrt(dof*M_PI)));
+
+    const float upperLimit(15.f);
+    const float dx((upperLimit-tTestStatistic)/static_cast<float>(nIntegrationSteps));
+    float integral(tDistCoeff*std::pow( 1.f + tTestStatistic*tTestStatistic/dof, -0.5f*(dof + 1.f)) + 
+            tDistCoeff*std::pow( 1.f + upperLimit*upperLimit/dof, -0.5f*(dof + 1.f)));
+    for (unsigned int iStep = 1; iStep < nIntegrationSteps; iStep++)
+        integral+=2.f*tDistCoeff*std::pow( 
+            1.f + (tTestStatistic + static_cast<float>(iStep)*dx)*(tTestStatistic + static_cast<float>(iStep)*dx)/dof, -0.5f*(dof + 1.f));
+    integral *= dx/2.f;
+
+    return integral;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+template <typename T>
+float LArDiscreteProbabilityHelper::CalculateCorrelationCoefficient(const T &t1, const T &t2)
+{
+    if (LArDiscreteProbabilityHelper::GetSize(t1) != (LArDiscreteProbabilityHelper::GetSize(t2)))
+        throw pandora::StatusCodeException(pandora::STATUS_CODE_INVALID_PARAMETER);
+
+    if (2 > LArDiscreteProbabilityHelper::GetSize(t1))
+        throw pandora::StatusCodeException(pandora::STATUS_CODE_INVALID_PARAMETER);
+
+    const float mean1(LArDiscreteProbabilityHelper::CalculateMean(t1));
+    const float mean2(LArDiscreteProbabilityHelper::CalculateMean(t2));
+
+    float variance1(0.f), variance2(0.f), covariance(0.f);
+
+    for (size_t iElement = 0; iElement < LArDiscreteProbabilityHelper::GetSize(t1); iElement++)
+    {
+        float element1(LArDiscreteProbabilityHelper::GetElement(t1,iElement));
+        float element2(LArDiscreteProbabilityHelper::GetElement(t2,iElement));
+
+        variance1 += (element1-mean1)*(element1-mean1);
+        variance2 += (element2-mean2)*(element2-mean2);
+        covariance += (element1-mean1)*(element2-mean2);
```

Comment:
I went for reducing the line numbers and making it look a bit more like maths on a piece of paper.  I don't have strong feelings on this one though. 

A good middle ground is to store the diff rather than the element, I think.
So, fixed!